### PR TITLE
demo and doc for sparse training with local and cluster model

### PR DESCRIPTION
**Pull request by backyes**
1. add QuickStart for sparse training
2. add Performance Analysis on Sparse Training
3. add Internals for Sparse training
4. add some FAQ 
@reyoung

### CODE REVIEW

File: doc/demo/sparse_train/sparse_train.md [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    --saving_period=1 \
+    --test_period=0 \
+    --num_passes=4 \
+    --dot_period=2 \
+    --log_period=20 \
+    --trainer_count=10 \
+    --saving_period_by_batches=5000 \
+    --local=1
+```
+For cluster train, you maybe just care about ```--ports_num_for_sparse=4``` in command line and conf.py files.  You need to understand simple cluster helper scripts in ```paddle/cluster_train/paddle.py```. Check DOC about cluster training.
```
Comment (backyes) [inline doc/demo/sparse_train/sparse_train.md]: 4 is necessary for current experiments. If it's is too little, setParameters API could generate BIG protobuf request body that is not recommended. In addition, BIG size can benefit communication performance for little size cluster JOB.

File: doc/demo/sparse_train/sparse_train.md [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+        # give data to paddle.
+        yield sparse_non_values, label
+    f.close()  # close file
+```
+Please consult related DOC for ```sparse_binary_vector``` and ```integer_value```.
+
+### Prepare Model Configuration
+Here providing one simple fc_layers network, which is enough for explaining HOW to train DNN with sparse training.
+
+We use ```ParameterAttribute(sparse_update=True)``` to enable sparse training for local job as well as cluster job. The ```trainer_config_helpers``` model will set internal FLAGs automatically. At last, these layers not set with sparse FLAG do dense training while sparse layers automatically use sparse training. Generally RECOMMEND you set these layers with high sparsity input, such as first hidden layers with sparse input.
```
Comment (backyes) [inline doc/demo/sparse_train/sparse_train.md]: This detail can help user understand what happened in internally, which is prefered in this DOC. We hope users understand that the Sparse Training is not only several sparse Matrix muliplication, but also involving forwardbackward,optimisation, some additional design for cluseter.

File: demo/sparse_train/sparse_binary/sparse_trainer_config.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+from paddle.trainer_config_helpers import *
+
+label_size = 2
+data_size = 18182296
+
+
+""" Algorithm Configuration """
+settings(learning_rate=1e-3,
+         learning_method=MomentumOptimizer(momentum=0.9),
```
Comment (backyes) [inline demo/sparse_train/sparse_binary/sparse_trainer_config.py]: done

File: demo/sparse_train/sparse_binary/sparse_trainer_config.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+         batch_size=200)
+
+""" Data Configuration """
+define_py_data_sources2(train_list='train.list',
+                        test_list=None,
+                        module='sparse_data_provider',
+                        obj='process')
+
+""" Model Configuration """
+non_value = data_layer(name='data',
```
Comment (backyes) [inline demo/sparse_train/sparse_binary/sparse_trainer_config.py]: done

File: doc/demo/sparse_train/sparse_train.md [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+## Quick Start
+
+Assuming that you are familiar with PaddlePaddle. Here show train billion level parameters with sparse training. 
+
+### Prepare Data
+Given the Input data sample as follows:
+```
+0,0,7216737 1,18127654 1, ... 
+```
+The first slot ```0``` is just one flag, not one feature; the second slot ```0``` is  label; the third slot is feature id, ```7216737 1```means the ```7216737``` dimension is not zero, the ```1``` is just one flag, not one feature; the ```18127654 1``` is another non zero feature. The line data depicts one simple high dimension input data for classification task.
```
Comment (backyes) [inline doc/demo/sparse_train/sparse_train.md]: the 'slot' here just is used to describe this sample, not related PaddlePaddle. I will add further emphasis on Dataprovider usage.

File: demo/sparse_train/sparse_vector/sparse_float_trainer_config.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+         batch_size=200)
+
+""" Data Configuration """
+define_py_data_sources2(train_list='train.list',
+                        test_list=None,
+                        module='sparse_float_data_provider',
+                        obj='process')
+
+""" Model Configuration """
+sparse_float = data_layer(name='data',
```
Comment (backyes) [inline demo/sparse_train/sparse_vector/sparse_float_trainer_config.py]: done

File: doc/demo/sparse_train/sparse_train.md [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+
+All samples are stored as simple TXT file ```train.txt```, then set it in ```train.list``` index file, which is used in model configuration file. We just provides no more than 2000+ samples to explains following  experiments. Check related HOWTO in other DOC.
```
Comment (backyes) [inline doc/demo/sparse_train/sparse_train.md]: zip it to <500KB, the size maybe ok.

File: doc/demo/sparse_train/sparse_train.md [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+```
+0,0,7216737 1,18127654 1, ... 
+```
+The first slot ```0``` is just one flag, not one feature; the second slot ```0``` is  label; the third slot is feature id, ```7216737 1```means the ```7216737``` dimension is not zero, the ```1``` is just one flag, not one feature; the ```18127654 1``` is another non zero feature. The line data depicts one simple high dimension input data for classification task.
+
+All samples are stored as simple TXT file ```train.txt```, then set it in ```train.list``` index file, which is used in model configuration file. We just provides no more than 2000+ samples to explains following  experiments. Check related HOWTO in other DOC. 
+
+### Prepare DataProvider
+Use following dataprovider to feed DataProvider:
+```python
```
Comment (backyes) [inline doc/demo/sparse_train/sparse_train.md]: done

File: doc/demo/sparse_train/sparse_train.md [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        splits.pop(0)
+        sparse_non_values = []
+        for value in splits:
+            v = value.split(" ")
+            sparse_non_values.append(long(v[0]))
+        # give data to paddle.
+        yield sparse_non_values, label
+    f.close()  # close file
+```
+Please consult related DOC for ```sparse_binary_vector``` and ```integer_value```.
```
Comment (backyes) [inline doc/demo/sparse_train/sparse_train.md]: done

File: doc/demo/sparse_train/sparse_train.md [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+
+hidden1 = fc_layer(input=non_value,
+                   size=128,
+                   param_attr=ParameterAttribute(sparse_update=True))
+hidden2 = fc_layer(input=hidden1,
+                   size=32)
+
+prediction = fc_layer(input=hidden2, size=label_size, act=SoftmaxActivation())
+
+outputs(classification_cost(input=prediction, label=label))
```
Comment (backyes) [inline doc/demo/sparse_train/sparse_train.md]: done

**Comment by backyes** [comment]
follow most comments

**Comment by backyes** [comment]
Goto: https://github.com/baidu/Paddle/pull/172
